### PR TITLE
3dview properties

### LINKED ISSUE #4154: [FEATURE][3D] Billboard Rendering for Points (#31308)
Original commit: https://github.com/qgis/QGIS/commit/b9978d53d11fc205207f98df8d749a1b8e3445be by wonder-sk

* Add button to change billboard symbol.

* Add shaders for billboards.

* Add billboard material class.

* Add sample billboard image.

* Add dummy handler for billboard.

* Add billboard geometry class.

* First working version of 3D point as billboard.

* Make selected feature bigger.

* Use default symbol preview as billboard.

* Update billboard when changes the viewport size.

* Use QgsTerrainTextureImage for adding image to texture.

* Show better quality of billboard default symbol.

* Promote QgsSymbolButton not working.

* Add QgsSymbolButton properly.

* Add seTexture2DFromSymbol.

* Use default symbol

* [On Progress] Get symbol from the QgsSymbolButton.

* [On progress] failed to change the symbol

* Fix billboard fragment shader for macOS

* Better transparency handling.

* Use size from the symbol.

* Add set layer for 3d billboard symbol.

* Set billboard size from the symbol pixel size * 20

* Use size from the image.

* [Debug] set white symbol to success kid, it work when previous billboard is success kid.

* Use store and restore symbol from Nyall.

* Remove unrelated code.

* Apply random extent to fix the bug, from Martin.

* Remove add mesh entities for billboard.

* Hide material and transformation widget for billboard.

* Remove add scene entities for billboard.

* Add selected as parameter in drawPreviewIcon.

* Add QgsMarkerSymbol to store the symbol in billboard

* Put billboard symbol in point3dsymbol serialization.

* Add billboard height in the UI.

* Set height for billboard.

* Store billboard height separately. Perhaps better to use existing transform.

* Set height of billboard to TY.

* Flipped image for billboard symbol.

* Add helper function to set billboard function.

* Generate billboard transform from general transform.

* Use static dpi for context size.

* Remove debug message.

* Use Qgs3DRenderContext for selection color and dpi from QgsApplication.

* Remove unnecessary file and code.

* Remove commented iut code.

* Add documentation of new classes.

* Fix banned keywords.

* Add missing parameter's documentation.

* Address PR review.

* Use context from readXML.

* Remove one line function.

* Fix billboard size.

* Make some functions private, remove unused function.

* Use unique_ptr for the ownership.

* Use unique_ptr for default symbol to avoid memory leak.

* Add parent to texture2D.

* Fix documentation.

* Add copy constructor.

* Fix documentation.

* Remove documentation warning.

* Fix sip error.

* Add unit test for billboard rendering.

* Set ambient to not transparent for measurement line.

* Add dpi in the Qgs3DMapSetting

* Update billboard rendering unit test.

* Remove layer from 3d point widget since it's not supported yet.

* Addressing PR review from Martin.

* Attach dom directly for the billboard symbol.

* Remove unused constructor for point3dsymbol.

* Use stroke width ratio for solving big stroke width.

* Update unit test for rendering with some stroke color.

* Fix typo.

### PR DESCRIPTION
**Pull request by lajunek**
Goal: Show users a short approach to visualize 3d vectors in QGIS. We need to add the helsinki.qgz
to the QGIS sample dataset.

Ticket(s): fix #4154 

- [x ] Backport to LTR documentation is required

### CODE REVIEW

**Comment by lajunek** [comment]
@saberraz @havatv @DelazJ @SrNetoChan can you have a look at this approach?

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 1]
```diff
@@ -1845,6 +1845,80 @@ rendering:
...
+          (see https://en.wikipedia.org/wiki/Phong_reflection_model#Description)
+	* If you choose a :guilabel:`3D Model` a whole model is depicted at the position
+          of single point coordinates.
+	* For e.g. visualizing point clouds you can use :guilabel:`Billboard` Shapes
+          defined by the :guilabel:`Billboard Height`, :guilabel:`Billboard symbol`and
+          :guilabel:`Altitude clamping`. The symbol will have a stable size.
+	* The :guilabel:`Altitude clamping` by :guilabel:`Absolute`, :guilabel:`Relative`
+	  or :guilabel:`Terrain`. These settings refer to the 3D data you also
+	  loaded into the 3D view (raster elevation data or 3D vector data) and 
+	  define the location for the anchor point of the 3D shape.
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]: This is not easy to explain what this means, but perhaps something like

```suggestion
	  define how z/elevation values are interpreted with respect to the underlying terrain surface.
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 2]
```diff
@@ -1845,6 +1845,80 @@ rendering:
...
+	* The :guilabel:`Translation` can be used to move objects in x, y and z axis.
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
	* :guilabel:`Translation` can be used to move objects in x, y and z axis.
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 3]
```diff
@@ -1845,6 +1845,80 @@ rendering:
...
+------------
+	
+	* You can define different simple 3D shapes like :guilabel:`Sphere`, :guilabel:`Cylinder´,
+          :guilabel:`Cube`, :guilabel:`Cone`, :guilabel:`Plane` and :guilabel:`Torus`
+          defined by their :guilabel:`Radius`, :guilabel:`Size` or :guilabel:`Length´.
+	  The unit of size of the 3D shapes refers to the CRS of the project.
+	* The shading of the 3D shapes can be defined by the menus :guilabel:`Diffuse`, 
+	  :guilabel:`Ambient`, :guilabel:`Specular` and :guilabel:`Shininess`
+          (see https://en.wikipedia.org/wiki/Phong_reflection_model#Description)
+	* If you choose a :guilabel:`3D Model` a whole model is depicted at the position
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
	* If you choose :guilabel:`3D Model`, their location will be determined
	  by the points in the layer.
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 4]
```diff
@@ -1845,6 +1845,80 @@ rendering:
...
+          of single point coordinates.
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 5]
```diff
@@ -1845,6 +1845,80 @@ rendering:
...
+	* For e.g. visualizing point clouds you can use :guilabel:`Billboard` Shapes
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
	* For visualizing 3D point clouds you can use :guilabel:`Billboard` Shapes
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 6]
```diff
@@ -1845,6 +1845,80 @@ rendering:
...
+          defined by the :guilabel:`Billboard Height`, :guilabel:`Billboard symbol`and
+          :guilabel:`Altitude clamping`. The symbol will have a stable size.
+	* The :guilabel:`Altitude clamping` by :guilabel:`Absolute`, :guilabel:`Relative`
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
	*  :guilabel:`Altitude clamping` can be set to :guilabel:`Absolute`, :guilabel:`Relative`
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 7]
```diff
@@ -1845,6 +1845,80 @@ rendering:
...
+	  or :guilabel:`Terrain`. These settings refer to the 3D data you also
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
	  or :guilabel:`Terrain`. These settings refer to the terrain surface defined by the 3D data you
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 8]
```diff
@@ -1845,6 +1845,80 @@ rendering:
...
+	  loaded into the 3D view (raster elevation data or 3D vector data) and
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
	  loaded into the 3D view (raster elevation data or 3D vector data), and 
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 9]
```diff
@@ -1845,6 +1845,80 @@ rendering:
...
+	* You can define a :guilabel:`Scale factor` for the 3D shape as well as a
+	  :guilabel:`Rotation` around the x-, y- and z-axis.
+
+
+Line layers
+-----------
+	
+	* Beneath the :guilabel:`Width`and :guilabel:`Height` settings you can
+	  define the :guilabel:`Extrusion` of the vector lines. If there are no
+	  z-values this way it is possible to define the height of the vectors in
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
	  z-values, this makes it is possible to define the height of the lines in
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 10]
```diff
@@ -1845,6 +1845,80 @@ rendering:
...
+	* The :guilabel:`Translation` can be used to move objects in x, y and z axis.
+	* You can define a :guilabel:`Scale factor` for the 3D shape as well as a
+	  :guilabel:`Rotation` around the x-, y- and z-axis.
+
+
+Line layers
+-----------
+	
+	* Beneath the :guilabel:`Width`and :guilabel:`Height` settings you can
+	  define the :guilabel:`Extrusion` of the vector lines. If there are no
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
	  define the :guilabel:`Extrusion` of the vector lines. If the lines do not have
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 11]
```diff
@@ -1845,6 +1845,80 @@ rendering:
...
+	  z-values this way it is possible to define the height of the vectors in
+          3D.
+	* With the :guilabel:`Altitude clamping` you define the position of the 
+	  3D lines on a surface, given that you inserted raster elevation data or
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
	  3D lines relative to the underlying terrain surface, if you have included raster elevation data or
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 12]
```diff
@@ -1845,6 +1845,80 @@ rendering:
...
+          other 3D vectors. We recommend to use :guilabel:`Absolute` here.
+	* The :guilabel:`Altitude binding` defines how the feature is clamped to the
+	  terrain. Either every :guilabel:`Vertex` of the feature will be clamped
+          to the terrain or this will be done by the :guilabel:`Centroid`.
+	* It is possible to |checkbox|:guilabel:`Render as simple 3D lines`.
+	* Also here the shading of the 3D lines can be defined by the menus
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
	* Also for lines, the shading can be defined in the menus
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 13]
```diff
@@ -1845,6 +1845,80 @@ rendering:
...
+          :guilabel:`Diffuse`, :guilabel:`Ambient`, :guilabel:`Specular` and
+          :guilabel:`Shininess`.
+
+Polygon Layers
+--------------
+	
+	* Here, also the :guilabel:`Height` can be defined in CRS units.
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
	* As for the other ones, :guilabel:`Height` can be defined in CRS units.
```

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 14]
```diff
@@ -1845,6 +1845,80 @@ rendering:
...
+
+|3d| The :guilabel:`3D View` tab provides settings for vector layers that should be
+depicted in the :ref:`label_3dmapview` tool. There are settings available for point, line
+and polygon layers. We provide information for the |singleSymbol|:guilabel:`Single symbol`
+rendering here.
+
+Point Layers
+------------
+	
+	* You can define different simple 3D shapes like :guilabel:`Sphere`, :guilabel:`Cylinder´,
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]: Travis is complaining here. You should avoid tabs, and make sure that the indentation is consistent. There are several more of these indentation "errors" further on in the text.

**Comment by lajunek** [comment]
@havatv removed the tabs but Travis is still complaining

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 15]
```diff
@@ -1845,6 +1845,77 @@ rendering:
...
+
+|3d| The :guilabel:`3D View` tab provides settings for vector layers that should be
+depicted in the :ref:`label_3dmapview` tool. There are settings available for point, line
+and polygon layers. We provide information for the |singleSymbol|:guilabel:`Single symbol`
+rendering here.
+
+Point Layers
+------------
+	
+* You can define different simple 3D shapes like :guilabel:`Sphere`, :guilabel:`Cylinder´,
```
Comment (havatv) [inline source/docs/user_manual/working_with_vector/vector_properties.rst]:
```suggestion
* You can define different simple 3D shapes like :guilabel:`Sphere`, :guilabel:`Cylinder`,
```

**Comment by havatv** [comment]
> @havatv removed the tabs but Travis is still complaining

In the suggestion i just provided, I have fixed the issue that caused Travis to complain.

**Comment by lajunek** [comment]
@hatav sorry. If I want to go back to the intendation you chose: how should I create an intendation? With spaces and how many.

**Comment by havatv** [comment]
> @hatav sorry. If I want to go back to the intendation you chose: how should I create an intendation? With spaces and how many.

Use spaces. The number of spaces depends on the context.

Could you accept / commit my suggestion from 30 minutes ago (changing a hyphen (') to an accent (`)), so we can see if Travis will be completely happy?

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 16]
```diff
@@ -1845,6 +1845,77 @@ rendering:
...
+  as explained above.
+* There is an additional option to |checkbox|:guilabel:`Add back faces`
+  and |checkbox|:guilabel:`Invert normals`.
+* You can define |checkbox|:guilabel:`Edges` by :guilabel:`Width` and :guilabel:`Color`.
+
+Application example
+-------------------
+
+To go through the settings explained above you can have a look at the qgis sample data
+and open the project :file:`helsinki.qgz`.
```
[the remark this answers is not in the document]
Comment (lajunek) [reply source/docs/user_manual/working_with_vector/vector_properties.rst]: @DelazJ my idea would be to extend the qgis sample dataset with 3d vectors files and a raster with elevation data from the Helsinki dataset I downloaded from Sabers Dropbox. The data are very suitable for setting up a 3d view.

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 17]
```diff
@@ -1845,6 +1845,77 @@ rendering:
...
+  to the terrain or this will be done by the :guilabel:`Centroid`.
+* It is possible to |checkbox|:guilabel:`Render as simple 3D lines`.
+* Also for lines, the shading can be defined in the menus :guilabel:`Diffuse`, :guilabel:`Ambient`,
+  :guilabel:`Specular` and :guilabel:`Shininess`.
+
+Polygon Layers
+--------------
+	
+* As for the other ones, :guilabel:`Height` can be defined in CRS units.
+* Again, :guilabel:`Extrusion` is possible for missing z-values.
```
[the remark this answers is not in the document]
Comment (lajunek) [reply source/docs/user_manual/working_with_vector/vector_properties.rst]: @saberraz can we go into detail here. When do we need to set the Extrusion parameter.Thanks(-:

File: source/docs/user_manual/working_with_vector/vector_properties.rst [hunk 18]
```diff
@@ -1845,6 +1845,77 @@ rendering:
 
 See :ref:`data_defined_labeling` for more information.
 
+.. index:: 3d view properties
+.. _`sec_3_d_view`:
+
+|3d| 3D View Properties
```
[the remark this answers is not in the document]
Comment (lajunek) [reply source/docs/user_manual/working_with_vector/vector_properties.rst]: @DelazJ yes we do have icons

[the remark this answers is not in the document]
Comment (DelazJ) [reply source/docs/user_manual/working_with_vector/vector_properties.rst]: I meant there's no icon in the section title (https://docs.qgis.org/testing/en/docs/user_manual/working_with_vector/vector_properties.html)

File: source/docs/user_manual/working_with_vector/vector_properties.rst (hunk 16 above)
[the remark this answers is not in the document]
Comment (DelazJ) [reply source/docs/user_manual/working_with_vector/vector_properties.rst]: I agree that having 3d data in the sample dataset will be helpful when testing 3D features, describing them and screen shooting. An already set project file in the docs repository would also be of great help.
My concern relates to adding the project file to the sample dataset, hence accessible to any user. Such a project could get outdated in few releases and no one would notice. I think that explaining settings (and providing data) should be enough to help anyone to check how features do work, in their own project.

[the remark this answers is not in the document]
Comment (lajunek) [reply source/docs/user_manual/working_with_vector/vector_properties.rst]: @DelazJ there is a link to a ed scene now. Can somebody replace and insert the screenshot I uploaded here? I would prefer a screenshot from ubuntu in this section, which I'm not using. Thanks a lot !

[the remark this answers is not in the document]
Comment (havatv) [reply source/docs/user_manual/working_with_vector/vector_properties.rst]: > @DelazJ there is a link to a ed scene now. Can somebody replace and insert the screenshot I uploaded here? I would prefer a screenshot from ubuntu in this section, which I'm not using. Thanks a lot !

3d_view_properties

**Comment by DelazJ** [comment]
@saberraz may I?

**Comment by DelazJ** [comment]
I think we should move on so I squashed/rebased and added a screenshot compatible with 3.10 so that we can backport.

**Comment by havatv** [comment]
Done.

**Pull request merged**